Add a `galois_group()` function

skieffer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


#### Brief description of what is fixed or changed

Add a `galois_group()` function

#### Other comments

Polynomials up through degree 6 are supported. Higher degrees could be attempted at a later time.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
  * Add a `galois_group()` function, supporting polynomials up through degree 6.
<!-- END RELEASE NOTES -->

skieffer committed:
Add a `galois_group()` function

oscarbenjamin commented on sympy/polys/numberfields/tests/test_galoisgroups.py:
```diff
@@ -0,0 +1,107 @@
+            L = QQ.alg_field_from_poly(U)
+            assert field_isomorphism(K.ext, L.ext) is not None
+        except MaxTriesException:
+            print('Max tries exceeded on Tschirnhausen transf.')
```
If this is only catching an exception from the first line then only the first line should be in the try block.

oscarbenjamin replied:
Also test code shouldn't print anything. It should either raise or pass silently.

skieffer replied:
I eliminated the non-determinism, so the `try-except` is gone.

oscarbenjamin commented on sympy/polys/numberfields/galoisgroups.py:
```diff
@@ -0,0 +1,808 @@
+
+def is_square(n):
+    """Convenience to check if an Integer is a square. """
+    return isinstance(sqrt(n), Integer)
```
```
from sympy.ntheory.primetest import is_square
```

skieffer replied:
Thanks -- thought that ought to already exist somewhere!

oscarbenjamin commented:
I guess this is incomplete but it broadly looks good.

Would it make sense for `galois_group` to be a method on `Poly` e.g. `Poly(...).galois_group()`?

skieffer committed:
Use existing `is_square()` function

skieffer committed:
Remove unnecessary `try-except`

Transformations are explored in fixed order, so we don't have to worry about max tries
in this unit test.

sylee957 commented on sympy/polys/numberfields/galoisgroups.py:
```diff
@@ -0,0 +1,808 @@
+    ...
+
+
+class Resolvent:
```
Do we need to have a new class to solve this problem?
I'd like the usage of objects to be minimal as possible. If it can be replaced by functions and compositions of primitive types (like list, dict, tuple), it is often trivial to define the whole object for adhoc purposes, and make the library more inflexible to use, unless you have defined some calculus by polymorphism.

sylee957 reviewed (commented):
I'd try to do [annotation](https://docs.python.org/3/library/typing.html) for the every types of functions, intermediate variables such that it can also pass mypy. 
It can give an overview of how the library functions are cleanly typed or not.

jksuom commented:
This looks good to me. The computation of the resolvent is fairly involved and depends on several parameters, hence I think that collecting those in a class is justified.

skieffer committed:
Checkout new vers from dev branch

skieffer committed:
Cherry-pick from master branch

skieffer commented:
> try to do annotation for the every types of functions, intermediate variables such that it can also pass mypy.

I'd like to hold off on this sort of thing until after the whole discussion #17945 has been resolved.

skieffer commented:
> Would it make sense for galois_group to be a method on Poly e.g. Poly(...).galois_group()?

Yes, it would! I've added this, but called it `Poly(...).gal()`.

I also added an `AlgebraicField.gal()` method.

skieffer commented:
Okay, I've implemented up to degree 6, and that's as much as I want to do, for now.


# Guide for reviewers


## Things to ignore

I had to cherry-pick some recent changes from the master branch to get unit tests to pass.
Hence, can ignore:

* combinatorics/homomorphisms.py
* polys/rings.py


## Main focus


* polys/numberfields/*: We add the `galois_group()` function, plus supporting tools for computing resolvents.
* polys/polytools.py: make a `Poly.gal()` method.
* polys/domains/algebraicfield.py: make an `AlgebraicField.gal()` method.
* combinatorics/galois.py: defines the return values for the `galois_group()` function.


## Other

* combinatorics/named_groups.py: some refactoring in support of `combinatorics/galois.py`
* core/evalf.py: supports root approximation in `Resolvent` class
* polys/rootoftools.py: helps make root approximation in `Resolvent` class *faster* (by about five times!)


## Implementations not used by the `galois_group()` function

The module `polys/numberfields/galoisgroups.py` defines several functions, each of which computes Galois groups for a single degree, and uses some combination of techniques, indicated by its name:

* `_galois_group_degree_4_root_approx()`
* `_galois_group_degree_4_lookup()`
* `_galois_group_degree_5_hybrid()`
* `_galois_group_degree_5_lookup_ext_factor()`
* `_galois_group_degree_6_lookup()`

Of these, only the fastest ones -- those with `lookup` in their name -- are used by `galois_group()`.

So why do the others exist? Because it's not clear how well the lookup approach scales. Eventually, I or someone else might want to extend support to degrees 7, 8, 9, 10, ..., and the lookup table computations can very quickly become intractable for new and different resolvents.

The root approx method therefore might be the one that's needed, going forward (or perhaps some hybrid of the two). So it might be nice to keep it around. Initially, root approx was taking 4.5 minutes to complete the test suite, but I got that down to 52s by using a better approximation method. (I used `CRootOf.eval_approx()` instead of `evalf(CRootOf)`.) With more improvements to interval refinement, as we have discussed before (e.g. [here](https://github.com/sympy/sympy/pull/22737#issuecomment-1000592349)), this might become even more viable.

Meanwhile, there are other, more sophisticated methods out there, involving e.g. p-adics. But I think what we have here is a nice enough start, and does a fair bit without requiring any of that machinery.

skieffer commented:
The change to `setup.cfg` is to make `flake8` ignore the module where the lookup table is defined.
Otherwise we get

```
"pyflakes[F]" failed during execution due to RecursionError('maximum recursion depth exceeded')
```

oscarbenjamin commented:
This should be reported as a bug somewhere:

> Otherwise we get
> 
> ```
> "pyflakes[F]" failed during execution due to RecursionError('maximum recursion depth exceeded')
> ```

I just took a look at the diff and there is a *lot* here. There is no way I can verify the changes here in detail but I'll try to take a general look soon.

sylee957 commented on sympy/combinatorics/galois.py:
```diff
@@ -0,0 +1,589 @@
+from sympy.combinatorics.permutations import Permutation
+
+
+class S1TransitiveSubgroups:
```
It could better be something like `Enum`

skieffer replied:
Good point, and I was wondering if an actual `Enum` should be used. I didn't find any existing ones in SymPy so far, but I guess it's fine to start now.

skieffer replied:
Using `Enum` now. I think it's an improvement.

sylee957 commented on sympy/combinatorics/galois.py:
```diff
@@ -0,0 +1,589 @@
+    )
+
+
+def M20():
```
I think that `snake_case` naming should be encouraged.

asmeurer replied:
I would make an exception when it names a mathematical value written with a specific convention, like this presumably is.

asmeurer replied:
Also an important coding convention to use in SymPy is that when an algorithm is written from a book or paper, as is the case here, one should always try to match the algorithm to the paper with things like variable names and so on. That makes it much easier to cross-reference the book or paper in the future if there is ever some issue.

sylee957 replied:
it is less important for denotational problem because snakecase-camelcase transformation is just capitalizing the words over it

sylee957 replied:
And I assume that hyphens in things like `S3_in_S6` are just a workaround for spacing than to match exactly the original denotation. So that argument neither wouldn't strictly follow.

skieffer replied:
I do feel that mathematical conventions are more important than Python conventions in cases like these (although I understand the desire to have a uniform codebase).

For a different but related example, PEP8 recommends that variables always be lowercase, but that would be unthinkable in SymPy. (How could, say, a group and its element not be `G` and `g`?)

sylee957 replied:
I think that capital with hyphens could closely fit to `CONSTANT_CASE` (for example, `S3_IN_S6`) 
if you are happy with it.

sylee957 commented on sympy/combinatorics/galois.py:
```diff
@@ -0,0 +1,589 @@
+    return G
+
+
+def find_transitive_subgroups_of_S6(*targets, verbose=True):
```
I don't think that printing functions should be turned on by default,
It could be better to get rid of debug for production.

asmeurer replied:
You can use the debug functionality to only print things with SYMPY_DEBUG is set.

skieffer replied:
It defaults to False now. I didn't use the debug switch, since I didn't intend this for debugging, but for reproducing the functions that return fixed representations.

sylee957 commented on sympy/polys/numberfields/galois_resolvents.py:
```diff
@@ -0,0 +1,667 @@
+    return [ZZ(1)] + [c(*T_coeffs) for c in L]
+
+
+if __name__ == "__main__":
+    import sys
+    verbose = '-v' in sys.argv[1:]
+    trial_run = '-t' in sys.argv[1:]
+    table = generate_lambda_lookup(verbose=verbose, trial_run=trial_run)
+    print(table)
```
This debug script should be removed.

skieffer replied:
Ah, for debug scripts I would definitely agree, but this is actually functional! This is so that

    (.venv) $ python -m sympy.polys.numberfields.galois_resolvents

is a reproducable way to generate the `resolvent_lookup.py` module. How about if I add a comment to that effect, right above this.

skieffer replied:
I've added the comment.

skieffer commented:
> there is a lot here

True. I'd be happy to break this up into a series of smaller PRs.

oscarbenjamin commented:
> break this up into a series of smaller PRs.

I don't think that's necessary. I was just cautioning that it would take me some time to look at this in any detail. It looks like others can look at it more quickly though so I guess that's not a problem.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -3870,6 +3870,30 @@ def cancel(f, g, include=False):
         else:
             return tuple(map(per, result))
 
+    def gal(f, by_name=False, max_tries=30, randomize=False):
```
I think that the name could be 'galois'. That is not too long.

skieffer replied:
Sure. I was only thinking of `gal()` because of the notation $Gal(E/F)$. But I suppose `galois()` might be more easily discoverable by new users.

skieffer replied:
Also a reasonable perspective. I'm honestly fine with any of these options. I guess tab-completion in IDEs means shortness is really a non-issue.

oscarbenjamin replied:
There are many precedents for `Poly` methods having the exact same name as sympy functions that operate on Expr:
```
In [1]: d = {}

In [2]: exec('from sympy import *', d)

In [3]: len(d)
Out[3]: 913

In [4]: d.keys() & set(dir(Poly))
Out[4]: 
{'LC',
 'LM',
 'LT',
 'cancel',
 'cofactors',
 'compose',
 'content',
 'count_ops',
 'count_roots',
 'decompose',
 'degree',
 'degree_list',
 'diff',
 'discriminant',
 'div',
 'exquo',
 'factor_list',
 'gcd',
 'gcdex',
 'gff_list',
 'ground_roots',
 'half_gcdex',
 'homogeneous_order',
 'integrate',
 'intervals',
 'invert',
 'lcm',
 'monic',
 'nroots',
 'nth_power_roots_poly',
 'pdiv',
 'per',
 'pexquo',
 'pquo',
 'prem',
 'primitive',
 'quo',
 'real_roots',
 'refine',
 'refine_root',
 'rem',
 'resultant',
 'root',
 'simplify',
 'sqf_list',
 'sqf_norm',
 'sqf_part',
 'sturm',
 'subresultants',
 'terms_gcd',
 'total_degree',
 'trunc'}
```
Not all of those are equivalent but most are.

skieffer replied:
Interesting. Maybe the `galois_group` function should be made to fit into the existing pattern as well.

skieffer replied:
I've adapted it to fit the general pattern. The main implementation is now in the `Poly` method. The top-level function accepts `Expr`, converts to `Poly`, and uses the `Poly` method.

skieffer replied:
Also the method names in `Poly` and `AlgebraicField` are now `galois_group`, matching the top-level function.

jksuom commented:
I like this method of deriving resolvent polynomials by specializing generic resolvent polynomials. I was surprised to see how big some of the coefficients are in resolvent_lookup. However, it seems that the lambdas are evaluated as efficiently as possible.

skieffer committed:
Comments and typos

skieffer committed:
Turn off default printing

The kwarg is also renamed, since its purpose was never to support debugging, but
to support regenerating the various functions in this module that return a fixed
representation of each of several named groups.

skieffer committed:
Add notes explaning origin of fixed group reps

skieffer committed:
Use proper `Enum`s for transitive subgroups

skieffer committed:
Rename methods `gal` --> `galois_group`

skieffer committed:
Make `galois_group()` accept `Expr`

skieffer committed:
Repair docs

skieffer commented:
> how big some of the coefficients are

I agree, and I was also surprised that they seem to exhibit no obvious pattern, considering how much symmetry is involved.

skieffer commented on sympy/polys/numberfields/galoisgroups.py:
```diff
@@ -0,0 +1,622 @@
+
+
+@public
+def galois_group(f, gens=None, args=None,
+                 by_name=False, max_tries=30, randomize=False):
```
I wanted to ask advice on this. I'm trying to make the `galois_group` function fit into the pattern of functions that accept an `Expr`, convert it to a `Poly`, and then operate on that. One example would be,

https://github.com/sympy/sympy/blob/c4d6b3eb1c7b39832116f7b02baf524a1ffadfba/sympy/polys/polytools.py#L5839-L5843

There are many such functions, and they all seem to use the same signature, `(f, *gens, **args)`. Should I do the same?

What I have at the moment is kind of a mishmash. It makes `gens` and `args` into keyword args themselves. If I wanted to fit into the `(f, *gens, **args)` pattern, I would have to move the three special keyword args `by_name=False, max_tries=30, randomize=False` into `**args`.

I guess I'm struggling with this because I kind of see "two worlds" colliding here: there's the world of functions using the `(f, *gens, **args)` signature -- which I think are probably on the older side -- and there's the world of functions that are trying to follow the (newer?) rules for writing docstrings, in particular, [the rules about the parameters section](https://docs.sympy.org/latest/contributing/docstring.html#parameters-section).

Do we need rules for documenting `**kwargs` dictionaries?

In a case like this, I would want to document the three kwargs `by_name=False, max_tries=30, randomize=False` that are special to this function, and then somehow say that, "anything else in `**args` will be passed onward to the `poly_from_expr()` function; see that function for further details." I don't know if there's a protocol for this yet. I know there has been some discussion (#23376) about trying to avoid use of `**kwargs` where possible.

skieffer replied:
Yeah, good question, and I had similar thoughts about the poly options -- they don't serve much purpose here.

Maybe I can just omit `*gens` and `**args` completely? If the user passes a constant expression, that's just going to result in an error anyway, so I don't see any need for `*gens`.

oscarbenjamin replied:
For a constant polynomial would it not be the trivial group?

skieffer replied:
A case could probably be made either way. What's for sure is that it's one of those funny, degenerate cases. I think my thinking was something like, "What's the group of all permutations of an empty set of roots? Should be an empty group, since there aren't any ways to permute no roots?"

Alternatively, there's "one way," to permute no roots, namely, to "do nothing," which is what the identity permutation does. : )

I guess in practical terms it doesn't matter much.

skieffer replied:
Yeah, going back to the actual definition instead of trying to reason about "permutations of nothing" is certainly a good approach! And I did think about it that way as well, but wound up asking myself whether constant polynomials are regarded as even having splitting fields. I guess it may depend on the author. My source of truth on this stuff has always been Gallian's algebra, and when I went back there the answer was no.

Maybe the key is in the word "splitting," which refers to factorization, so it's my fault for introducing the subject of roots at all. As you point out, we should be thinking about polynomials over a field $K$. In that case, in all questions about factoring and irreducibility, constant polynomials are either zero, or units in the ring $K[x]$, and thus not even the sorts of things for which factorization is a question.

Gallian's system seems like a sensible one: $f(x) \in K[x]$ *splits* in an extension $L/K$ if it can be factored as a product of linear factors in $L[x]$. $L$ is a splitting field for $f(x)$ over $K$ if $f(x)$ splits in $L$ but in no proper subfield of $L$.

oscarbenjamin commented:
Is there a reason to reject polynomials with rational coefficients?

skieffer commented:
I dug into a thesis by L. Soicher, https://spectrum.library.concordia.ca/id/eprint/4916/1/MK49693.pdf, which is cited by Cohen, and it yielded some nice things.

---------------------------------------------------------------------------

> Is there a reason to reject polynomials with rational coefficients?

No, we can handle these. I'll push an update.

Soicher reminds me there's a nice trick for always having an algebraic integer as your primitive element, even if you started
with a non-integer. Given monic $f(x) \in \mathbb{Q}[x]$, replace this with $c^n f(x/c)$, where $n = deg(f)$ and $c$ is the LCM of the coeff denominators. This is monic over $\mathbb{Z}[x]$ and just multiplies the roots by $c$, so not only do you get the same field, you get the same splitting field.

I should go back and use this with the integral basis function we did last year.

---------------------------------------------------------------------------

Soicher acknowledges the lookup table approach, and notes that tables have been published, going back to 1929! How did they compute them?

skieffer committed:
Redo signature of `galois_group()` function

skieffer committed:
Add `Poly.make_monic_over_integers_by_scaling_roots()` method

skieffer committed:
Make `galois_group()` accept polys over `QQ`

skieffer commented:
I couldn't find an existing operation to turn $f(x)$ into $c^n f(x/c)$ and ensure the result is over `ZZ`, so I gave `Poly` a method with the very clunky name, `make_monic_over_integers_by_scaling_roots()`. Happy to rename it something better.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3870,6 +3870,46 @@ def cancel(f, g, include=False):
+            n = f.degree()
+            fm = f.monic()
+            c, _ = fm.clear_denoms()
+            return (c**n * fm.compose(Poly(fm.gen / c))).to_ring(), c
```
There is the Poly.transform method for this step.

skieffer replied:
Perfect! Thanks!

skieffer committed:
Use `Poly.transform()`

jksuom commented on sympy/polys/rings.py:
```diff
@@ -552,6 +552,19 @@ def add_gens(self, symbols):
         return self.clone(symbols=list(syms))
 
+    def symmetric_poly(self, n):
+        """Return the symmetric poly of given degree over this ring's gens."""
```
elementary symmetric polynomial

skieffer replied:
Maybe I can save this for a subsequent PR?

I made the change, but it results in a merge conflict. That's because we currently have this situation:

1. I started the `galois-groups` topic branch.
2. I opened and merged PR #24541, which made improvements to rings.py (and introduced the sloppy docstring, which I agree should be corrected).
3. I copied the new rings.py into the `galois-group` topic branch, since it is needed here.

```
      o---o---o   <-- PR #24541 made changes to rings.py
     /         \
o---o---o---o---o---o master
 \               \
  \               \  <-- cherry-picked the new rings.py from the other PR, since it's needed here
   \               \
    o---o---o---o---o---o galois-groups
```

If I now resolve the competing line changes on GitHub, [the docs state](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github) that this will cause the entire `master` branch to be merged into the `galois-group` branch. I generally find those kinds of merges to be messy, and think it would be better not to do it.

skieffer replied:
Okay, the change is now made on master as well as here, hence no conflict.

skieffer committed:
Improve docstring

jksuom commented:
I think this is ready to go. The implementation looks good. I have not been able to verify everything in the degree 6 algorithm; the proof presented by Cohen consists only of a few hints. Still, there is no reason to doubt the algorithm.

oscarbenjamin commented:
Nice. Good work @skieffer and @jksuom !

oscarbenjamin commented:
Is there a way to construct the Galois group of a reducible polynomial from the groups of its irreducible factors? (I don't know if this is a very simple or a very hard thing to do)

Also does knowing the group make it possible to compute expressions for the roots? 
```
In [27]: p = x**5 - 5*x + 12

In [28]: roots(p)
Out[28]: {}

In [29]: galois_group(p)
Out[29]: 
(PermutationGroup([
     (0 1 2 3 4),
     (0 4)(1 3)]),
 True)
```
Here this is solvable but SymPy can't solve it.

skieffer commented:
Thanks very much to everyone for reviewing!

The other questions are interesting. Will have to do some digging...

For now I can only say that, in the 1981 thesis by Soicher mentioned above, they didn't yet have a method for the reducible case (see p. 11).

oscarbenjamin commented on sympy/combinatorics/galois.py:
```diff
@@ -0,0 +1,611 @@
+
+    def get_perm_group(self):
+        if self == self.C4:
+            return CyclicGroup(4)
```
This gives a DeprecationWarning under Python 3.12 alpha 5:
```
      File "/home/runner/work/sympy/sympy/sympy/combinatorics/galois.py", line 83, in get_perm_group
        if self == self.C4:
                   ^^^^^^^
      File "/opt/hostedtoolcache/Python/3.12.0-alpha.5/x64/lib/python3.12/enum.py", line 210, in __get__
        warnings.warn(
    DeprecationWarning: `member.member` access (e.g. `Color.RED.BLUE`) is deprecated and will be removed in 3.14.
```
This causes the doctests to fail in #24776.

We need to get this fixed in both master and also on the 1.12 branch that I have just created for the release.
